micro_g/scripts/pose_servoing_controller.py

```diff
@@ -0,0 +1,194 @@
+        self.rate = self.core.create_rate(control_update_rate)
+
+        # Define the transforms we'll use. Notation is T_{to}{from}.
+        # Frames are:
+        #   - s = space frame, fixed to the robot's base link
+        #   - y = yaw frame, same as s but rotated about z by the current waist angle
+        #   - b = body frame, fixed to the end effector
+        #   - d = desired frame for the end effector
+        self.T_sy = np.identity(4)
+        self.T_sd = self.arm.get_ee_pose_command()
+        self.update_frames()
+
+        # Initialize the tf2 buffer and listener
+        self.tf_buffer = tf2_ros.Buffer()
+        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self.core)
+        # Set the target frame for transformation
+        self.target_frame = f"{robot_name}/base_link"
+
+        # Create a subscription to the pose that we will track
+        self.core.create_subscription(
+            PoseStamped, "desired_eef_pose", self.desired_pose_callback, 10
+        )
+        time.sleep(0.5)
+        self.core.get_logger().info("Ready to receive servoing commands.")
+
+    def update_frames(self):
+        """
+        Update the coordinate frames used by this controller.
+
+        We need to do this manually because the y frame (base frame rotated to match
+        arm yaw) is not actually published as a TF frame.
+        """
+        T_sb = self.arm.get_ee_pose_command()  # Current EEF pose in space frame
+        rpy = ang.rotation_matrix_to_euler_angles(T_sb[:3, :3])
+        self.T_sy[:2, :2] = ang.yaw_to_rotation_matrix(rpy[2])
+
+    def desired_pose_callback(self, msg: PoseStamped):
+        """
+        Process a desired pose
+
+        Args:
+            msg: A PoseStamped message
+        """
+        # Transform the pose into the base link frame
+        try:
+            # Lookup the transform from the source frame to target frame
+            transform = self.tf_buffer.lookup_transform(
+                self.target_frame, msg.header.frame_id, msg.header.stamp
+            )
+
+            # Transform the pose using the received transform
+            transformed_pose = do_transform_pose(msg.pose, transform)
+            position = transformed_pose.position
+            position = np.array([position.x, position.y, position.z])
+            orientation = transformed_pose.orientation
+            orientation = transforms3d.quaternions.quat2mat(
+                [orientation.w, orientation.x, orientation.y, orientation.z]
+            )
+
+            # Save the position and orientation into the homogeneous transform
+            self.T_sd[:3, :3] = orientation
+            self.T_sd[:3, 3] = position
+            self.core.get_logger().info(f"Received new desired pose T_sd:\n{self.T_sd}")
+
+        except Exception as e:
+            self.core.get_logger().error(
+                f"Error transforming pose to {self.target_frame}: {e}"
+            )
+            raise (e)
+
+    def controller(self):
```

Comment:
Maybe rename to something like `run_controller`. ros2_control likes to use `update`

Reply:
Done!